Wa/adding unit test

wellavelino opened the pull request:
Adding tests for gem commands to ensure the correct scaffolding.

cs-wellington-santos committed:
Adding rspec tests for gem commands

cs-wellington-santos committed:
Remove unnecessary code

cs-wellington-santos committed:
Adding new tests

- Adding tests to generate independent feature
- Adding tests to generate independent feature with invalid argument
- Adding tests to generate independent feature with alias

cs-wellington-santos committed:
New organization for specs

- Deleted hello_test.rb
- Adding .rspec file to configure output
- Change Rakefile
- Adding context to specs

cs-wellington-santos committed:
Adding simplecov

- Adding simplecov
- Adding corevage directory to .gitignore

cs-wellington-santos committed:
Adding new tests and improviments

- Include Rake task to run testes
- Include .rspec file configuration
- Adding simplecov has dependency

cs-wellington-santos committed:
Adding Circle CI

- Include circle.yml configuration

cs-wellington-santos committed:
Adding new configuration to circle.yml

cs-wellington-santos committed:
Adjust circle.yml to continue integration

cs-wellington-santos committed:
Adding 'gem install' into circle.yml

cs-wellington-santos committed:
Adding bundle install

cs-wellington-santos committed:
Remove simplecov from project

- Remove all simplecov dependencies
- Create Rake tasks 'spec_tests'

oscartanner commented:
Hi!
Thank you for your PR. Can you provide a description?
Best regards

cs-wellington-santos committed:
Refactor configuration files

- Changing gemspec tests file to test|spec|features
- Remove unecessary line in zip_helper.rb
- Changing Rake task name to ":tests"

oscartanner commented:
I executed `bundle exec rake spec_tests` and all the outputs from the gem are shown between the `rspec` outputs. It would be nicer if the gem outputs could be redirect to a file ([example here](https://codedump.io/share/lplqfkm4aHwn/1/suppress-console-output-during-rspec-tests))

wellavelino commented:
@oscartanner, Hi.

NP,  im adding a description for the pull request.

wellavelino commented:
@oscartanner, im adding a solution for this problem `system "command", :out => File::NULL`  to supress this console output.

cs-wellington-santos committed:
Adding suppress console output

-Adding suppress output for gem commands
-Adding output file into .gitignore

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+      end
+
+
+      it 'Should return the latest version of sunonomo using alias command' do
```
The alias `suno` is made with 

```
require 'sunomono'
Sunomono::SunomonoRunner.start
```

So if `sunomono` works, `suno` works too. I think that only one test with the alias is OK to ensure that it works

oscartanner replied:
The alias `suno` is made with 

```
require 'sunomono'
Sunomono::SunomonoRunner.start
```

So if `sunomono` works, `suno` works too. I think that only one test with the alias is OK to ensure that it works

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+        version = "#{Sunomono::VERSION}"
+
+        expect(version).
+            to end_with('0.2.0.pre')
```
This is no good.

If we change the gem version this test will fail. I think it is better to compare the return of the `system` command with the `Sunomono::VERSION` value

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+            to include('.', '..', 'build_app.rb', 'build_app.yml', 'devices.txt', 'run_tests_all_devices.sh')
+      end
+
+      it 'Create new project using alias command' do
```
There are a lot of duplicated code here.

I think it would be better if you choose another function to test the alias or put all these verifications together.

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+    end
+
+    context 'Using I18n to translate files' do
+      it 'Generate an OS independent feature using I18n to portuguese language' do
```
What about 'Generate an OS independent feature in Portuguese'?

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+      end
+    end
+
+    context 'This command should generates an OS independent feature(all files)' do
```
I think that would be better to change the order and organization here. This context is not so different from the above.

What about:

```
context  'Generate Feature'
it 'Generate all files'
it 'Generate in pt'
it 'Generate in en'
```

The idea is to use each command as one context

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,431 @@
+        Dir.chdir('../')
+
+        expect(Dir.entries("#{@project_name}/features")).
+            to include('.', '..', 'android', 'ios', 'step_definitions', 'support', "#{feature_name}.feature")
```
All these folder are not created by the g feature command and you tested it's creation on the first it.
I don see why re-check it here

cs-wellington-santos committed:
New improvements to sunomono tests

- Adding improviment to dont print output in terminal
- Getting gem version from output terminal
- New organize
- Change the names of contexts to make sense

cs-wellington-santos committed:
Changing circle ci configuration

wellavelino commented:
@oscartanner all the issues are ok now!

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,366 @@
+  describe 'Sunomono gem commands' do
+    context 'Returns project created with all files' do
+      it 'Create new project using default command' do
+        system "sunomono new '#{@project_name}'", :out => File::NULL
```
Try redirect to /dev/null and tell me what happens. If that don't work, try:

```
system "sunomono new '#{@project_name}' > /dev/null"
```

You don't need to create a new file on every execution.

cs-wellington-santos replied:
Nice Catch!

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,366 @@
+            to include("######### DADO #########\n", "\n", "######### QUANDO #########\n", "\n", "######### ENTãO #########")
+      end
+
+      it 'Generate feature in en' do
```
I thought about something. What if we use `I18n.available_locales = [:pt, :en, ...]` in the gem?

We could check all locales in one `it` using a `loop`

cs-wellington-santos replied:
ts a nice idea, we can add other languages more easily and also validate.

oscartanner replied:
Exactly. Maybe for this PR that would be too much. But we can think to add it in the future.

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,366 @@
+      it 'Generate feature in pt' do
+        system "suno new '#{@project_name}'", :out => File::NULL
+
+        feature_name = 'sunomono'
```
You use this variable in a lot of `it` with two different values `sunomono` and `rspec_sunomono`.
Any special reason for this?

cs-wellington-santos replied:
Its because i use rspec_sunomono in feature name

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,366 @@
+
+        Dir.chdir(@project_name)
+
+        system "sunomono generate feature '#{feature_name}' invalid_argument", :out => File::NULL
```
It would be nice to see if there was an error message.

cs-wellington-santos replied:
I think this is not a good idea, in others tests like unit this verification make more sense.

oscartanner replied:
I am not talking about the role message, but to check if there was an "Error" word in the console output, exactly the same way you do with `sunomono --version` test

oscartanner commented on spec/sunomono_test.rb:
```diff
@@ -0,0 +1,366 @@
+
+        system "suno generate feature '#{feature_name}' invalid_argument", :out => File::NULL
+
+        Dir.chdir('../')
```
You are already inside the project folder, why do you return to its parent?
With this, you need to add the `@project_name` in all `expect` sentences below.

cs-wellington-santos replied:
Yes, nice catch!

oscartanner replied:
This is what I was talking about:

```
context 'Generate Feature' do
  it 'Generates all files' do
    system "sunomono new '#{@project_name}'", :out => File::NULL

    Dir.chdir(@project_name)

    feature_name = 'rspec_sunomono'

    system "sunomono generate feature '#{feature_name}'", :out => File::NULL

    expect(Dir.entries("features")).
      to include("#{feature_name}.feature")
    expect(Dir.entries("features/step_definitions")).
      to include("#{feature_name}_steps.rb")
    expect(Dir.entries("features/android/screens")).
      to include("#{feature_name}_screen.rb")
    expect(Dir.entries("features/ios/screens")).
      to include("#{feature_name}_screen.rb")
end
```

oscartanner replied:
:)

cs-wellington-santos committed:
Refactor tests cases

- Remove several unecessary code
- New improviment to get gem version

oscartanner reviewed (approved):
👍
